src/Modules/ModulesGraded.jl

```diff
@@ -26,62 +32,394 @@ Graded free module Multivariate Polynomial Ring in x, y over Rational Field grad
+  x -> [1]
+  y -> [1], MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x, y])
+
+julia> S, _ = quo(R, [x*y])
+(Quotient of Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1] by ideal(x*y), Map from
+Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1] to Quotient of Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1] by ideal(x*y) defined by a julia-function with inverse)
+
+julia> F = free_module(S, 2)
+Free module of rank 2 over Quotient of Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1] by ideal(x*y)
+
+julia> FF = grade(F, [1, 2])
+Graded free module Quotient of Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1] by ideal(x*y)^1([-1]) + Quotient of Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1] by ideal(x*y)^1([-2]) of rank 2 over Quotient of Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1] by ideal(x*y)
+```
+"""
+function grade(F::FreeMod, W::Vector{<:Vector{<:IntegerUnion}})
+  @assert length(W) == ngens(F)
+  R = base_ring(F)
+  @assert is_zm_graded(R)
+  n = length(W[1])
+  @assert all(x->length(x) == n, W)
+  A = grading_group(R)
+  return grade(F, [A(w) for w = W])
+end
+
+function grade(F::FreeMod, W::Union{ZZMatrix, Matrix{<:IntegerUnion}})
+  @assert size(W, 2) == ngens(F)
+  R = base_ring(F)
+  @assert is_zm_graded(R)
+  A = grading_group(R)
+  return grade(F, [A(W[:, i]) for i = 1:size(W, 2)])
+end
+
+function grade(F::FreeMod, W::Vector{<:IntegerUnion})
+  @assert length(W) == ngens(F)
+  R = base_ring(F)
+  @assert is_z_graded(R)
   A = grading_group(R)
   N = free_module(R, length(W))
   N.d = [W[i] * A[1] for i in 1:length(W)]
-  N.S = M.S
+  N.S = F.S
   return N
 end
 
+@doc raw"""
+    grading_group(F::FreeMod)
+
+Return the grading group of `base_ring(F)`.
+
+# Examples
+```jldoctest
+julia> R, (x,y) = graded_polynomial_ring(QQ, ["x", "y"]);
+
+julia> F = graded_free_module(R, 3)
+Graded free module Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1]^3([0]) of rank 3 over Multivariate Polynomial Ring in x, y over Rational Field graded by
+  x -> [1]
+  y -> [1]
+
+julia> grading_group(F)
+GrpAb: Z
+```
+"""
+function grading_group(M::FreeMod)
+  return grading_group(base_ring(M))
+end
+
+
+# Dangereous: Only for internal use with care!!!
+@doc raw"""
+    set_grading!(F::FreeMod, W::Vector{GrpAbFinGenElem})
+
+    set_grading!(F::FreeMod, W::Vector{<:Vector{<:IntegerUnion}})
+
+    set_grading!(F::FreeMod, W::Union{ZZMatrix, Matrix{<:IntegerUnion}}) 
+
+    set_grading!(F::FreeMod, W::Vector{<:IntegerUnion})
+
+Assign weights to the generators of `F` according to the entries of `W`.
+
+See the `grade` and `graded_free_module` functions.
+```
```

Comment:
Maybe you also want to put a warning in this docstring and actively encourage the use of grade and graded_free_module instead of set_grading!  ?

Reply:
Yes, note however that set_grading! will not be added to the docu and is only for internal use with care. It already has a warning comment.